README.md
# PDF Table Extractor

Simple tool to extract tables from PDF files.

## Installation
```bash
pip install tabula-py
```

## Usage
```python
from pdf_extractor import extract_pdf
tables = extract_pdf("document.pdf")
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.0.0] - 2025-11-26

### Added
- Multi-language support for French, German, and English documents
- Automatic locale detection from filename patterns
- Content-based locale detection using keyword analysis
- Locale-aware date parsing (DD/MM/YYYY, DD.MM.YYYY, MM/DD/YYYY)
- Locale-aware amount parsing (different decimal/thousands separators)
- CLI options: `--locale {fr,de,en}` and `--no-auto-detect`
- Python API: `default_locale`, `auto_detect_locale`, `explicit_locale` parameters
- Comprehensive locale detection tests
- Multi-strategy PDF table extraction (lattice, stream, guess)
- Bank statement parser with transaction extraction
- CSV and Markdown output formats
- Batch directory processing
- Command-line interface (`pdf-extract`)
- Python API for programmatic use
- GitHub Actions CI/CD pipeline
- PyPI package configuration
- Comprehensive documentation
- MIT License
- Contributing guidelines

### Locale Support Details
- **French (fr)**: Default locale
  - Date format: DD/MM/YYYY or DD.MM.YYYY
  - Amount format: 1 234,56 € (space thousands, comma decimal)
  - Keywords: solde, crédit, débit, virement, prélèvement, etc.

- **German (de)**:
  - Date format: DD.MM.YYYY
  - Amount format: 1.234,56 € (dot thousands, comma decimal)
  - Keywords: saldo, kredit, lastschrift, überweisung, betrag, etc.

- **English (en)**:
  - Date format: MM/DD/YYYY or YYYY-MM-DD
  - Amount format: 1,234.56 $ (comma thousands, dot decimal)
  - Keywords: balance, credit, debit, transfer, payment, etc.

### Detection Priority
1. Explicit `--locale` parameter
2. Filename patterns (_fr, _de, _en, français, deutsch, english, etc.)
3. Content analysis (keyword matching with confidence threshold)
4. Default locale (French)

## [Unreleased]

### Planned Features
- Additional parsers for invoices, receipts, and forms
- More language support (Spanish, Italian, Portuguese, etc.)
- OCR support for scanned PDFs
- Excel and JSON output formats
- Web UI (Flask/Streamlit)
- Docker image for easy deployment

---

**Note**: This is the initial release of PDF AI Context Extractor. The project was created to provide clean, structured PDF data extraction optimized for AI agent consumption, with a focus on French administrative documents but supporting multiple languages.


tests/test_extractor.py
"""Tests for PDF extractor."""
import pytest
from pathlib import Path
from pdf_ai_context_extractor import PDFTableExtractor

def test_extractor_initialization():
    """Test extractor can be initialized."""
    extractor = PDFTableExtractor(verbose=False)
    assert extractor is not None

def test_extractor_verbose_mode():
    """Test verbose mode setting."""
    extractor = PDFTableExtractor(verbose=True)
    assert extractor.verbose is True


tests/generate_samples.py
#!/usr/bin/env python3
"""
Generate anonymized sample PDFs for testing.

Creates realistic but anonymized sample documents that can be safely
shared in the GitHub repository.
"""

from reportlab.lib.pagesizes import letter, A4
from reportlab.lib import colors
from reportlab.platypus import SimpleDocTemplate, Table, TableStyle, Paragraph, Spacer
from reportlab.lib.styles import getSampleStyleSheet
from reportlab.lib.units import inch
from pathlib import Path
import random
from datetime import datetime, timedelta


def generate_bank_statement_pdf(output_path: Path):
    """Generate a sample bank statement PDF."""

    doc = SimpleDocTemplate(str(output_path), pagesize=A4)
    elements = []
    styles = getSampleStyleSheet()

    # Title
    title = Paragraph("<b>BANK STATEMENT - Sample Account</b>", styles['Title'])
    elements.append(title)
    elements.append(Spacer(1, 0.2*inch))

    # Account info
    info_data = [
        ['Account Holder:', 'John DOE'],
        ['Account Number:', '****1234'],
        ['Period:', 'January 2024'],
        ['Balance:', '$2,345.67'],
    ]
    info_table = Table(info_data, colWidths=[2*inch, 3*inch])
    info_table.setStyle(TableStyle([
        ('FONTNAME', (0, 0), (-1, -1), 'Helvetica'),
        ('FONTSIZE', (0, 0), (-1, -1), 10),
        ('BACKGROUND', (0, 0), (0, -1), colors.lightgrey),
    ]))
    elements.append(info_table)
    elements.append(Spacer(1, 0.3*inch))

    # Transactions table
    transactions_data = [
        ['Date', 'Description', 'Amount', 'Balance']
    ]

    # Generate sample transactions
    start_date = datetime(2024, 1, 1)
    balance = 2500.0

    sample_transactions = [
        ('Salary Deposit', 3000.00),
        ('Supermarket Purchase', -85.50),
        ('Utility Bill - Electricity', -120.00),
        ('Online Purchase', -45.99),
        ('ATM Withdrawal', -200.00),
        ('Restaurant', -67.80),
        ('Gas Station', -55.30),
        ('Insurance Premium', -95.00),
        ('Pharmacy', -28.45),
        ('Transfer to Savings', -500.00),
    ]

    for i, (desc, amount) in enumerate(sample_transactions):
        date = (start_date + timedelta(days=i*3)).strftime('%d/%m/%Y')
        balance += amount
        transactions_data.append([
            date,
            desc,
            f"${amount:.2f}",
            f"${balance:.2f}"
        ])

    trans_table = Table(transactions_data, colWidths=[1.2*inch, 3*inch, 1.2*inch, 1.2*inch])
    trans_table.setStyle(TableStyle([
        ('FONTNAME', (0, 0), (-1, 0), 'Helvetica-Bold'),
        ('FONTSIZE', (0, 0), (-1, -1), 9),
        ('BACKGROUND', (0, 0), (-1, 0), colors.grey),
        ('TEXTCOLOR', (0, 0), (-1, 0), colors.whitesmoke),
        ('ALIGN', (2, 0), (-1, -1), 'RIGHT'),
        ('GRID', (0, 0), (-1, -1), 0.5, colors.black),
        ('ROWBACKGROUNDS', (0, 1), (-1, -1), [colors.white, colors.lightgrey]),
    ]))
    elements.append(trans_table)

    # Build PDF
    doc.build(elements)
    print(f"✓ Generated: {output_path.name}")


def generate_invoice_pdf(output_path: Path):
    """Generate a sample invoice PDF."""

    doc = SimpleDocTemplate(str(output_path), pagesize=letter)
    elements = []
    styles = getSampleStyleSheet()

    # Title
    title = Paragraph("<b>INVOICE</b>", styles['Title'])
    elements.append(title)
    elements.append(Spacer(1, 0.2*inch))

    # Invoice info
    info_data = [
        ['Invoice #:', 'INV-2024-001'],
        ['Date:', '15/01/2024'],
        ['Customer:', 'ABC Corporation'],
        ['Payment Terms:', 'Net 30'],
    ]
    info_table = Table(info_data, colWidths=[1.5*inch, 3*inch])
    elements.append(info_table)
    elements.append(Spacer(1, 0.3*inch))

    # Line items
    items_data = [
        ['Item', 'Description', 'Qty', 'Price', 'Total']
    ]

    sample_items = [
        ('Consulting Services', 'Technical consulting - January', 20, 150.00),
        ('Software License', 'Annual subscription', 1, 500.00),
        ('Training Session', '2-day workshop', 2, 800.00),
    ]

    for item, desc, qty, price in sample_items:
        total = qty * price
        items_data.append([
            item,
            desc,
            str(qty),
            f"${price:.2f}",
            f"${total:.2f}"
        ])

    # Add totals
    subtotal = sum(qty * price for _, _, qty, price in sample_items)
    tax = subtotal * 0.10
    total = subtotal + tax

    items_data.extend([
        ['', '', '', 'Subtotal:', f"${subtotal:.2f}"],
        ['', '', '', 'Tax (10%):', f"${tax:.2f}"],
        ['', '', '', 'Total:', f"${total:.2f}"],
    ])

    items_table = Table(items_data, colWidths=[1.5*inch, 2.5*inch, 0.7*inch, 1*inch, 1*inch])
    items_table.setStyle(TableStyle([
        ('FONTNAME', (0, 0), (-1, 0), 'Helvetica-Bold'),
        ('BACKGROUND', (0, 0), (-1, 0), colors.grey),
        ('TEXTCOLOR', (0, 0), (-1, 0), colors.whitesmoke),
        ('ALIGN', (2, 0), (-1, -1), 'RIGHT'),
        ('GRID', (0, 0), (-1, -4), 0.5, colors.black),
        ('LINEABOVE', (3, -3), (-1, -3), 1, colors.black),
        ('LINEABOVE', (3, -1), (-1, -1), 2, colors.black),
    ]))
    elements.append(items_table)

    doc.build(elements)
    print(f"✓ Generated: {output_path.name}")


def generate_simple_table_pdf(output_path: Path):
    """Generate a simple data table PDF."""

    doc = SimpleDocTemplate(str(output_path), pagesize=letter)
    elements = []
    styles = getSampleStyleSheet()

    # Title
    title = Paragraph("<b>SAMPLE DATA TABLE</b>", styles['Title'])
    elements.append(title)
    elements.append(Spacer(1, 0.2*inch))

    # Simple table
    data = [
        ['ID', 'Name', 'Category', 'Value'],
        ['001', 'Item A', 'Type 1', '150.00'],
        ['002', 'Item B', 'Type 2', '275.50'],
        ['003', 'Item C', 'Type 1', '99.99'],
        ['004', 'Item D', 'Type 3', '450.00'],
        ['005', 'Item E', 'Type 2', '125.00'],
    ]

    table = Table(data, colWidths=[0.8*inch, 2*inch, 1.5*inch, 1.2*inch])
    table.setStyle(TableStyle([
        ('FONTNAME', (0, 0), (-1, 0), 'Helvetica-Bold'),
        ('BACKGROUND', (0, 0), (-1, 0), colors.lightblue),
        ('GRID', (0, 0), (-1, -1), 1, colors.black),
        ('ALIGN', (3, 1), (3, -1), 'RIGHT'),
    ]))
    elements.append(table)

    doc.build(elements)
    print(f"✓ Generated: {output_path.name}")


def main():
    """Generate all sample PDFs."""

    # Check if reportlab is installed
    try:
        import reportlab
    except ImportError:
        print("❌ reportlab not installed")
        print("   Install with: pip install reportlab")
        return

    samples_dir = Path(__file__).parent / "samples"
    samples_dir.mkdir(exist_ok=True)

    print("\n" + "="*80)
    print("GENERATING ANONYMIZED SAMPLE PDFs")
    print("="*80)
    print()

    # Generate samples
    generate_bank_statement_pdf(samples_dir / "sample_bank_statement.pdf")
    generate_invoice_pdf(samples_dir / "sample_invoice.pdf")
    generate_simple_table_pdf(samples_dir / "sample_data_table.pdf")

    print()
    print("="*80)
    print(f"✅ Generated 3 sample PDFs in: {samples_dir}")
    print("="*80)
    print()
    print("These anonymized PDFs can be safely committed to the repository")
    print("for testing and demonstration purposes.")
    print()


if __name__ == "__main__":
    main()


quick_setup.sh
#!/bin/bash
# Quick setup script for PDF AI Context Extractor development

set -e

echo "════════════════════════════════════════════════════════════════════"
echo "   PDF AI CONTEXT EXTRACTOR - Quick Setup"
echo "════════════════════════════════════════════════════════════════════"
echo ""

# Check Python
if ! command -v python3 &> /dev/null; then
    echo "❌ Python 3 not found. Please install Python 3.7+"
    exit 1
fi

echo "✓ Python found: $(python3 --version)"

# Check Java
if ! command -v java &> /dev/null; then
    echo "⚠️  Java not found. tabula-py requires Java Runtime Environment"
    echo "   Install with:"
    echo "   - Ubuntu/Debian: sudo apt install default-jre"
    echo "   - macOS: brew install java"
    exit 1
fi

echo "✓ Java found: $(java -version 2>&1 | head -n 1)"
echo ""

# Create virtual environment
echo "Creating virtual environment..."
python3 -m venv venv
source venv/bin/activate

# Install package in development mode
echo "Installing package in development mode..."
pip install -e ".[dev]"

# Install reportlab for sample generation
echo "Installing reportlab for test samples..."
pip install reportlab

# Generate test samples
echo ""
echo "Generating anonymized test PDFs..."
python tests/generate_samples.py

echo ""
echo "════════════════════════════════════════════════════════════════════"
echo "✅ Setup Complete!"
echo "════════════════════════════════════════════════════════════════════"
echo ""
echo "Next steps:"
echo "  1. Activate environment: source venv/bin/activate"
echo "  2. Run examples: python examples/run_examples.py"
echo "  3. Run tests: pytest"
echo ""
echo "To use the tool:"
echo "  pdf-extract path/to/document.pdf"
echo ""
echo "See README.md for full documentation"
echo ""


examples/run_examples.py
#!/usr/bin/env python3
"""
Example usage and test runner for PDF AI Context Extractor.

This script demonstrates the tool's capabilities and validates
that extraction works correctly on sample PDFs.
"""

from pathlib import Path
import sys

# Add src to path for development
sys.path.insert(0, str(Path(__file__).parent.parent / "src"))

from pdf_ai_context_extractor import PDFTableExtractor, bank_statement_parser


def run_examples():
    """Run extraction examples on sample PDFs."""

    print("="*80)
    print("PDF AI CONTEXT EXTRACTOR - EXAMPLES")
    print("="*80)
    print()

    # Find sample PDFs
    samples_dir = Path(__file__).parent.parent / "tests" / "samples"

    if not samples_dir.exists():
        print(f"❌ Samples directory not found: {samples_dir}")
        print("   Please create sample PDFs in tests/samples/")
        return False

    pdf_files = list(samples_dir.glob("*.pdf"))

    if not pdf_files:
        print(f"❌ No PDF files found in {samples_dir}")
        print("   Please add sample PDFs to tests/samples/")
        return False

    print(f"✓ Found {len(pdf_files)} sample PDF(s)")
    print()

    # Create output directory
    output_dir = Path(__file__).parent / "output"
    output_dir.mkdir(exist_ok=True)

    # Example 1: Basic extraction
    print("Example 1: Basic CSV Extraction")
    print("-" * 80)
    extractor = PDFTableExtractor(verbose=True)

    for pdf_file in pdf_files:
        print(f"\nProcessing: {pdf_file.name}")
        df, method = extractor.extract_tables(pdf_file)

        if df is not None:
            print(f"✓ Extracted {len(df)} rows using {method} method")
            print(f"  Columns: {', '.join(str(c) for c in df.columns[:5])}")
            if len(df.columns) > 5:
                print(f"  ... and {len(df.columns) - 5} more")

            # Save to CSV
            output_file = output_dir / f"{pdf_file.stem}_basic.csv"
            df.to_csv(output_file, index=False)
            print(f"  Saved to: {output_file}")
        else:
            print(f"✗ Failed to extract")

    print("\n" + "="*80)

    # Example 2: With bank statement parser
    print("\nExample 2: Bank Statement Parser")
    print("-" * 80)

    for pdf_file in pdf_files:
        print(f"\nProcessing: {pdf_file.name}")
        df, method = extractor.extract_tables(pdf_file)

        if df is not None:
            # Try applying bank statement parser
            parsed = bank_statement_parser(df, pdf_file.stem)

            if parsed is not None and not parsed.empty:
                print(f"✓ Parsed {len(parsed)} transactions")
                print(f"  Columns: {', '.join(parsed.columns)}")

                # Show sample
                print("\n  Sample transactions:")
                for idx, row in parsed.head(3).iterrows():
                    print(f"    {row['Date Operation']}: {row['Description'][:40]}... {row['Montant']:.2f}")

                # Save parsed
                output_file = output_dir / f"{pdf_file.stem}_parsed.csv"
                parsed.to_csv(output_file, index=False)
                print(f"\n  Saved to: {output_file}")
            else:
                print(f"  ⚠ Parser did not extract transactions (may not be bank statement format)")

    print("\n" + "="*80)

    # Example 3: Markdown output
    print("\nExample 3: Markdown Output for AI Prompts")
    print("-" * 80)

    for pdf_file in pdf_files[:1]:  # Just first file
        print(f"\nProcessing: {pdf_file.name}")
        df, method = extractor.extract_tables(pdf_file)

        if df is not None:
            # Limit to first 5 rows for demo
            sample = df.head(5)

            md_content = f"# Extracted from {pdf_file.stem}\n\n"
            md_content += sample.to_markdown(index=False)

            output_file = output_dir / f"{pdf_file.stem}_sample.md"
            output_file.write_text(md_content)

            print("✓ Created Markdown output:")
            print(md_content)
            print(f"\nSaved to: {output_file}")

    print("\n" + "="*80)
    print("\n✅ Examples complete!")
    print(f"   Output files in: {output_dir}")
    print()

    # Print summary
    extractor.print_summary()

    return True


def validate_installation():
    """Validate that all dependencies are installed."""
    print("Validating installation...")

    try:
        import tabula
        print("  ✓ tabula-py installed")
    except ImportError:
        print("  ✗ tabula-py not installed")
        print("    Install with: pip install tabula-py")
        return False

    try:
        import pandas
        print("  ✓ pandas installed")
    except ImportError:
        print("  ✗ pandas not installed")
        print("    Install with: pip install pandas")
        return False

    # Check Java
    import subprocess
    try:
        result = subprocess.run(['java', '-version'], capture_output=True, text=True)
        if result.returncode == 0:
            print("  ✓ Java installed")
        else:
            print("  ⚠ Java check returned non-zero")
    except FileNotFoundError:
        print("  ✗ Java not installed")
        print("    tabula-py requires Java Runtime Environment")
        return False

    print()
    return True


if __name__ == "__main__":
    print()
    if not validate_installation():
        print("\n❌ Installation validation failed")
        print("   Please install missing dependencies")
        sys.exit(1)

    print("✓ All dependencies installed\n")

    success = run_examples()

    sys.exit(0 if success else 1)


src/pdf_ai_context_extractor/cli.py
"""Command-line interface for PDF extractor."""
import argparse
import sys
from pathlib import Path
from .core import PDFTableExtractor
from .parsers import PARSERS

def main():
    parser = argparse.ArgumentParser(
        description='PDF AI Context Extractor',
        formatter_class=argparse.RawDescriptionHelpFormatter,
    )
    parser.add_argument('input', type=Path, help='PDF file or directory')
    parser.add_argument('-o', '--output', type=Path, help='Output directory')
    parser.add_argument('-f', '--format', choices=['csv', 'md', 'both'],
                       default='csv', help='Output format')
    parser.add_argument('-p', '--pattern', default='*.pdf',
                       help='Glob pattern for files')
    parser.add_argument('--parser', choices=list(PARSERS.keys()),
                       help='Custom parser')
    parser.add_argument('-q', '--quiet', action='store_true',
                       help='Suppress output')

    args = parser.parse_args()

    if not args.input.exists():
        print(f"Error: {args.input} does not exist")
        sys.exit(1)

    output_dir = args.output or args.input.parent / 'extracted'
    extractor = PDFTableExtractor(verbose=not args.quiet)

    if args.input.is_file():
        df, method = extractor.extract_tables(args.input)
        if df is not None:
            if args.format in ('csv', 'both'):
                df.to_csv(output_dir / f"{args.input.stem}.csv", index=False)
            if args.format in ('md', 'both'):
                extractor.save_as_markdown(df, output_dir / f"{args.input.stem}.md")
    else:
        extractor.process_directory(args.input, output_dir, args.pattern)

if __name__ == "__main__":
    main()


src/pdf_ai_context_extractor/core.py
"""Core PDF table extraction logic."""
import sys
from pathlib import Path
import pandas as pd

try:
    import tabula
except ImportError:
    print("ERROR: tabula-py not installed")
    sys.exit(1)

class PDFTableExtractor:
    """Multi-strategy PDF table extractor."""

    def __init__(self, verbose=True):
        self.verbose = verbose

    def extract_tables(self, pdf_path):
        """Extract tables from PDF."""
        if self.verbose:
            print(f"Processing: {pdf_path}")

        methods = ['lattice', 'stream', 'guess']
        best_result = None
        best_method = None
        max_rows = 0

        for method in methods:
            try:
                if method == 'lattice':
                    tables = tabula.read_pdf(str(pdf_path), pages='all', lattice=True)
                elif method == 'stream':
                    tables = tabula.read_pdf(str(pdf_path), pages='all', stream=True)
                else:
                    tables = tabula.read_pdf(str(pdf_path), pages='all', guess=True)

                if tables:
                    combined = pd.concat(tables, ignore_index=True)
                    if len(combined) > max_rows:
                        max_rows = len(combined)
                        best_result = combined
                        best_method = method
            except Exception:
                continue

        return best_result, best_method

    def save_as_markdown(self, df, output_path):
        """Save DataFrame as Markdown table."""
        md_content = f"# Extracted from {output_path.stem}\n\n"
        md_content += df.to_markdown(index=False)
        md_content += f"\n\n---\n*Extracted: {pd.Timestamp.now()}*\n"
        output_path.write_text(md_content, encoding='utf-8')

    def process_directory(self, input_dir, output_dir, pattern="*.pdf"):
        """Process all PDF files in a directory."""
        pdf_files = sorted(Path(input_dir).glob(pattern))

        if not pdf_files:
            print(f"No PDF files found in {input_dir}")
            return

        print(f"Found {len(pdf_files)} PDF files")
        Path(output_dir).mkdir(parents=True, exist_ok=True)

        for pdf_path in pdf_files:
            df, method = self.extract_tables(pdf_path)
            if df is not None:
                output_path = Path(output_dir) / pdf_path.stem
                df.to_csv(output_path.with_suffix('.csv'), index=False)
                print(f"  Processed: {pdf_path.name}")


src/pdf_ai_context_extractor/parsers.py
"""Custom parsers for specific PDF formats."""
import pandas as pd
import re

def bank_statement_parser(df, filename):
    """Parser for bank statement format."""
    transactions = []

    for idx, row in df.iterrows():
        row_str = ' '.join([str(x) for x in row if pd.notna(x)])

        # Look for dates (DD/MM/YYYY)
        dates = re.findall(r'\d{1,2}/\d{1,2}/\d{4}', row_str)
        # Look for amounts
        amounts = re.findall(r'[+-]?\s*\d+[,\s]\d{2}', row_str)

        if dates and amounts:
            date_op = dates[0]
            amount = amounts[-1].replace(' ', '').replace(',', '.')

            transactions.append({
                'Date': date_op,
                'Description': row_str,
                'Amount': float(amount),
                'Source': filename
            })

    if transactions:
        return pd.DataFrame(transactions)
    return None

PARSERS = {
    'bank_statement': bank_statement_parser,
}


src/pdf_ai_context_extractor/locale_detector.py
"""
Locale detection and configuration for multi-language PDF extraction.

Supports French (fr), German (de), and English (en) with automatic detection
from filename and content.
"""

import re
from typing import Optional, Tuple, Dict, Any
from pathlib import Path
import pandas as pd
from dataclasses import dataclass


@dataclass
class LocaleConfig:
    """Configuration for a specific locale."""

    code: str
    name: str
    date_patterns: list  # Regex patterns for date formats
    amount_patterns: list  # Regex patterns for amounts
    decimal_separator: str  # ',' or '.'
    thousands_separator: str  # ' ', '.', or ','
    keywords: list  # Common words in this language

    def __repr__(self):
        return f"LocaleConfig({self.code}, {self.name})"


# Locale configurations
LOCALES: Dict[str, LocaleConfig] = {
    'fr': LocaleConfig(
        code='fr',
        name='French',
        date_patterns=[
            r'\d{1,2}/\d{1,2}/\d{4}',  # DD/MM/YYYY
            r'\d{1,2}\.\d{1,2}\.\d{4}',  # DD.MM.YYYY
            r'\d{1,2}\s+(?:janvier|février|mars|avril|mai|juin|juillet|août|septembre|octobre|novembre|décembre)\s+\d{4}',
        ],
        amount_patterns=[
            r'[+-]?\s*\d+[\s,]\d{2}\s*€?',  # 1 234,56 or 1234,56
            r'[+-]?\s*\d+,\d{2}\s*EUR',
        ],
        decimal_separator=',',
        thousands_separator=' ',
        keywords=[
            'solde', 'crédit', 'débit', 'virement', 'prélèvement',
            'montant', 'date', 'opération', 'facture', 'relevé',
            'janvier', 'février', 'mars', 'avril', 'mai', 'juin',
            'juillet', 'août', 'septembre', 'octobre', 'novembre', 'décembre'
        ]
    ),

    'de': LocaleConfig(
        code='de',
        name='German',
        date_patterns=[
            r'\d{1,2}\.\d{1,2}\.\d{4}',  # DD.MM.YYYY
            r'\d{1,2}/\d{1,2}/\d{4}',  # DD/MM/YYYY
            r'\d{1,2}\s+(?:Januar|Februar|März|April|Mai|Juni|Juli|August|September|Oktober|November|Dezember)\s+\d{4}',
        ],
        amount_patterns=[
            r'[+-]?\s*\d+[\.,]\d{3}[\.,]\d{2}',  # 1.234,56 or 1,234.56
            r'[+-]?\s*\d+,\d{2}\s*€?',
            r'[+-]?\s*\d+,\d{2}\s*EUR',
        ],
        decimal_separator=',',
        thousands_separator='.',
        keywords=[
            'saldo', 'kredit', 'lastschrift', 'überweisung', 'betrag',
            'datum', 'buchung', 'rechnung', 'kontoauszug',
            'januar', 'februar', 'märz', 'april', 'mai', 'juni',
            'juli', 'august', 'september', 'oktober', 'november', 'dezember'
        ]
    ),

    'en': LocaleConfig(
        code='en',
        name='English',
        date_patterns=[
            r'\d{1,2}/\d{1,2}/\d{4}',  # MM/DD/YYYY or DD/MM/YYYY
            r'\d{4}-\d{1,2}-\d{1,2}',  # YYYY-MM-DD
            r'\d{1,2}\s+(?:January|February|March|April|May|June|July|August|September|October|November|December)\s+\d{4}',
        ],
        amount_patterns=[
            r'[+-]?\s*\$?\d+,\d{3}\.\d{2}',  # $1,234.56
            r'[+-]?\s*\d+\.\d{2}\s*(?:USD|EUR|GBP)?',
        ],
        decimal_separator='.',
        thousands_separator=',',
        keywords=[
            'balance', 'credit', 'debit', 'transfer', 'payment',
            'amount', 'date', 'transaction', 'invoice', 'statement',
            'january', 'february', 'march', 'april', 'may', 'june',
            'july', 'august', 'september', 'october', 'november', 'december'
        ]
    ),
}


class LocaleDetector:
    """Detect and manage locale for PDF extraction."""

    def __init__(self, default_locale: str = 'fr'):
        """
        Initialize locale detector.

        Args:
            default_locale: Default locale code ('fr', 'de', 'en')
        """
        if default_locale not in LOCALES:
            raise ValueError(f"Unknown locale: {default_locale}. Available: {list(LOCALES.keys())}")

        self.default_locale = default_locale
        self._confidence_threshold = 0.3  # Minimum confidence to detect locale

    def detect_from_filename(self, filename: str) -> Optional[str]:
        """
        Detect locale from filename patterns.

        Args:
            filename: PDF filename

        Returns:
            Detected locale code or None
        """
        filename_lower = filename.lower()

        # Language indicators in filename
        lang_indicators = {
            'fr': ['_fr', '-fr', 'french', 'francais', 'français', 'france'],
            'de': ['_de', '-de', 'german', 'deutsch', 'deutschland', 'germany'],
            'en': ['_en', '-en', 'english', 'uk', 'usa', 'us'],
        }

        for locale, indicators in lang_indicators.items():
            if any(ind in filename_lower for ind in indicators):
                return locale

        return None

    def detect_from_content(self, df: pd.DataFrame) -> Tuple[Optional[str], float]:
        """
        Detect locale from DataFrame content using keyword matching.

        Args:
            df: Extracted DataFrame

        Returns:
            Tuple of (detected locale code, confidence score)
        """
        if df is None or df.empty:
            return None, 0.0

        # Concatenate all text content
        text = ' '.join(
            str(val).lower()
            for row in df.values
            for val in row
            if pd.notna(val)
        )

        if not text:
            return None, 0.0

        # Count keyword matches for each locale
        scores = {}
        for locale_code, config in LOCALES.items():
            keyword_count = sum(
                1 for keyword in config.keywords
                if keyword.lower() in text
            )
            # Normalize by number of keywords
            scores[locale_code] = keyword_count / len(config.keywords)

        if not scores:
            return None, 0.0

        # Get best match
        best_locale = max(scores, key=scores.get)
        confidence = scores[best_locale]

        # Only return if confidence is above threshold
        if confidence >= self._confidence_threshold:
            return best_locale, confidence

        return None, confidence

    def detect_locale(
        self,
        filename: str,
        df: Optional[pd.DataFrame] = None,
        explicit_locale: Optional[str] = None
    ) -> Tuple[str, str]:
        """
        Detect locale using multiple strategies.

        Priority:
        1. Explicit locale parameter
        2. Filename pattern
        3. Content analysis
        4. Default locale

        Args:
            filename: PDF filename
            df: Extracted DataFrame (for content analysis)
            explicit_locale: Explicitly specified locale

        Returns:
            Tuple of (locale code, detection method)
        """
        # 1. Explicit locale
        if explicit_locale:
            if explicit_locale in LOCALES:
                return explicit_locale, 'explicit'
            else:
                print(f"⚠️  Unknown locale '{explicit_locale}', using default")

        # 2. Filename pattern
        filename_locale = self.detect_from_filename(filename)
        if filename_locale:
            return filename_locale, 'filename'

        # 3. Content analysis
        if df is not None:
            content_locale, confidence = self.detect_from_content(df)
            if content_locale:
                return content_locale, f'content (confidence: {confidence:.2%})'

        # 4. Default
        return self.default_locale, 'default'

    def get_config(self, locale: str) -> LocaleConfig:
        """Get configuration for a locale."""
        return LOCALES.get(locale, LOCALES[self.default_locale])


def clean_amount(
    text: str,
    locale: str = 'fr',
    locale_config: Optional[LocaleConfig] = None
) -> Optional[float]:
    """
    Clean and convert amount strings to float based on locale.

    Args:
        text: Amount text
        locale: Locale code
        locale_config: Optional LocaleConfig (avoids lookup)

    Returns:
        Float value or None
    """
    if pd.isna(text) or text == '':
        return None

    if locale_config is None:
        locale_config = LOCALES.get(locale, LOCALES['fr'])

    text = str(text).strip()

    # Remove currency symbols
    text = re.sub(r'[€$£¥]', '', text)
    text = re.sub(r'EUR|USD|GBP|CHF', '', text, flags=re.IGNORECASE)

    # Remove spaces
    text = text.replace(' ', '')

    # Handle locale-specific formatting
    if locale_config.decimal_separator == ',':
        # French/German: 1.234,56 or 1 234,56
        # Remove thousands separator
        text = text.replace(locale_config.thousands_separator, '')
        # Replace decimal separator
        text = text.replace(',', '.')
    else:
        # English: 1,234.56
        # Remove thousands separator
        text = text.replace(locale_config.thousands_separator, '')

    # Remove + signs
    text = text.replace('+', '')

    try:
        return float(text)
    except ValueError:
        return None


def parse_date(
    text: str,
    locale: str = 'fr',
    locale_config: Optional[LocaleConfig] = None
) -> Optional[str]:
    """
    Parse date string based on locale.

    Args:
        text: Date text
        locale: Locale code
        locale_config: Optional LocaleConfig

    Returns:
        Standardized date string (DD/MM/YYYY) or None
    """
    if pd.isna(text) or not text:
        return None

    if locale_config is None:
        locale_config = LOCALES.get(locale, LOCALES['fr'])

    text = str(text).strip()

    # Try each date pattern for this locale
    for pattern in locale_config.date_patterns:
        match = re.search(pattern, text)
        if match:
            return match.group(0)

    return None


# Convenience function for checking available locales
def list_available_locales() -> Dict[str, str]:
    """List all available locales."""
    return {code: config.name for code, config in LOCALES.items()}
